bundles/org.openhab.ui.basic/web-src/smarthome.js

```diff
@@ -1606,25 +1679,65 @@
 				_t.verify.cancel();
 			}
 
-			var newValue = value;
+			var newValue = cleanValue(value);
 			var undefValue = "";
-			if (itemState === "undefined" || itemState === "NULL" || itemState === "UNDEF") {
+			if (itemState === undefined) {
+				undefValue = lastUndef;
+			} else if (itemState === "NULL" || itemState === "UNDEF") {
+				if (_t.itemType === "datetime") {
+					undefValue = "";
+					if (_t.input.type === "text") {
+						undefValue = "YYYY-MM-DD hh:mm";
+					}
+				} else {
+					undefValue = !(newValue === "" || newValue === "NULL" || newValue === "UNDEF") ? newValue : lastUndef;
+				}
 				newValue = "";
-				undefValue = value;
 			}
 
+			if (_t.inputHint === "number") {
+				if (newValue !== "") {
+					newValue = parseNumber(newValue, _t.unit, true).value;
+					var valueArray = newValue.trim().split(" ");
+					newValue = valueArray[0];
+					if (valueArray.length > 1) {
+						_t.input.parentNode.nextElementSibling.innerHTML = valueArray[1];
+					}
+				} else {
+					var undefArray = undefValue.split(" ");
+					undefValue = undefArray[0];
+				}
+			} else if (_t.itemType === "datetime") {
+				newValue = ((itemState !== "NULL") && (itemState !== "UNDEF")) ? itemState : newValue;
+				newValue = newValue.trim().split(".")[0];		// drop millis
+				if (newValue.match(timeWithSecondsPattern)) {	// drop seconds
+					newValue = newValue.split(":").slice(0, -1).join(":");
+				}
+				if (newValue !== "") {
+					if (_t.inputHint === "date") {
+						newValue = newValue.split("T")[0];
+					} else if (_t.inputHint === "time") {
+						newValue = newValue.split("T")[1];
+					}
+					if (_t.input.type === "text") {
+						newValue = newValue.replace("T", " ");
+					}
+				}
+			}
 			_t.input.value = newValue;
 			_t.input.nextElementSibling.innerHTML = undefValue;
 
 			_t.input.parentNode.MaterialTextfield.change();
 			_t.input.parentNode.MaterialTextfield.checkValidity();
-
 			lastValue = value;
 			lastItemState = itemState;
```

Comment:
I am not sure if lastItemState should be updated when itemState is undefined (not passed to thge call) ?

Reply:
I am now sure it matters. But I notice itemState is not being set at all, and it should be set somewhere. That may explain the reverting to an old state.

Reply:
I believe lastItemState should set only if itemState is defined. That means it will be updated only when a new item state is received, not when you trigger a reset value to previous one because a wrong value was filled. 
Do you agree?

Reply:
I agree.